pkg/cmd/upgrade/upgrade_boot.go

```diff
@@ -0,0 +1,442 @@
+	if err != nil {
+		return errors.Wrapf(err, "failed to get boot config ref for version stream: %s", versionStreamRef)
+	}
+	upgradeSha, upgradeVersion, err := o.bootConfigRef(configCloneDir, versionStreamURL, upgradeVersionSha, bootConfigURL)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get boot config ref for version stream ref: %s", upgradeVersionSha)
+	}
+
+	// check if boot config upgrade available
+	if upgradeSha == currentSha {
+		log.Logger().Infof(util.ColorInfo("No boot config upgrade available"))
+		return nil
+	}
+	log.Logger().Infof(util.ColorInfo("boot config upgrade available"))
+	log.Logger().Infof("Upgrading from v%s to v%s", currentVersion, upgradeVersion)
+
+	err = o.Git().FetchBranch(o.Dir, bootConfigURL, "master")
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch master of %s", bootConfigURL)
+	}
+
+	err = o.cherryPickCommits(configCloneDir, currentSha, upgradeSha)
+	if err != nil {
+		return errors.Wrap(err, "failed to cherry pick upgrade commits")
+	}
+	err = o.excludeFiles(currentSha)
+	if err != nil {
+		return errors.Wrap(err, "failed to exclude files from commit")
+	}
+	return nil
+}
+
+func (o *UpgradeBootOptions) bootConfigRef(dir string, versionStreamURL string, versionStreamRef string, configURL string) (string, string, error) {
+	resolver, err := o.CreateVersionResolver(versionStreamURL, versionStreamRef)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to create version resolver %s", configURL)
+	}
+	configVersion, err := resolver.ResolveGitVersion(configURL)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to resolve config url %s", configURL)
+	}
+	cmtSha, err := o.Git().GetCommitPointedToByTag(dir, fmt.Sprintf("v%s", configVersion))
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to get commit pointed to by %s", cmtSha)
+	}
+	return cmtSha, configVersion, nil
+}
+
+func (o *UpgradeBootOptions) cloneBootConfig(configURL string) (string, error) {
+	cloneDir, err := ioutil.TempDir("", "")
+	err = os.MkdirAll(cloneDir, util.DefaultWritePermissions)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create directory: %s", cloneDir)
+	}
+
+	err = o.Git().CloneBare(cloneDir, configURL)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to clone git URL %s to directory %s", configURL, cloneDir)
+	}
+	return cloneDir, nil
+}
+
+func (o *UpgradeBootOptions) cherryPickCommits(cloneDir, fromSha, toSha string) error {
+	cmts := make([]gits.GitCommit, 0)
+	cmts, err := o.Git().GetCommits(cloneDir, fromSha, toSha)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get commits from %s", cloneDir)
+	}
+
+	log.Logger().Infof("cherry picking commits in the range %s..%s", fromSha, toSha)
+	for i := len(cmts) - 1; i >= 0; i-- {
+		commitSha := cmts[i].SHA
+		commitMsg := cmts[i].Subject()
+
+		err := o.Git().CherryPickTheirs(o.Dir, commitSha)
+		if err != nil {
```

Comment:
We need to be able to continue from where is left in case of a failure. Maybe we can write the last committed sha and the branch name in a file.

Reply:
Cerated https://github.com/jenkins-x/jx/issues/5582